### ISSUE TITLE
lsjson --hash on local filesystem is very inefficient

### ISSUE THREAD

**Issue by mattheys** [rclone/rclone#4181]
I was attempting to do an lsjson --hash on my local filesystem to produce MD5 hashes so I can compare with a lsjosn dump from Google Drive to see if I have duplicate files in different folders on Google Drive before uploading the files due to poor internet bandwidth.

https://forum.rclone.org/t/md5-hashing-local-filesystem-is-very-slow/15951

I noticed that it was incredibly slow to produce the hashes on larger files and even when doing lsf --hash MD5 it was still taking lots of time. (30mins vs 2mins with Powershell)

After reviewing the code I found that lsf calls lsjson and lsjson steps through and performs each hash sequentially on the file.

https://github.com/rclone/rclone/blob/8e91f83174cc4011158c596d0c8569cd226fc36d/fs/operations/lsjson.go#L155-L162

Would it be possible to rewrite this in a few ways, 

- One to pass a command line option to perform an array of hashes e.g. something like MD5,SHA256,SHA512 etc (from the list of already supported hashes) defaulting to all if not supplied
- Two to read the file once and perform all requested hashes concurrently to reduce disk access wear and IO bottle necks
- Three to perform the multiple hashes in a way where each one is on it's own thread to make use of multicore CPU's and reduce CPU as a bottle neck

Unfortunatly my Go skills will allow me to build a custom rclone binary that only does an MD5 hash but I don't think I would be able to add the above functionality to a pull request.

**Comment by ncw** [comment]
Yes you are right about this!

```
$ time md5sum 1G
cd573cfaace07e7949bc0c46028904ff  1G

real	0m2.160s
user	0m1.773s
sys	0m0.382s

$ time rclone hashsum MD5 1G
cd573cfaace07e7949bc0c46028904ff  1G

real	0m2.126s
user	0m2.039s
sys	0m0.454s

$ time rclone lsf --hash MD5 --format hp 1G
cd573cfaace07e7949bc0c46028904ff;1G

real	0m52.143s
user	0m50.705s
sys	0m3.831s
```

I had a go at fixing this here - what do you think?

https://beta.rclone.org/branch/v1.51.0-240-g2b570bd3-fix-4181-lsf-hash-beta/ (uploaded in 15-30 mins)

This introduces a `--hash-type` parameter to `rclone lsjson` which `rclone lsf` uses internally when `--hash` is specified.

**Comment by mattheys** [comment]
Thanks, just tested the Windows 64bit on Windows 10 and it worked really well. Completed the hash in the same time that Powershell did. --hash-type seems like the way to go.

**Comment by ncw** [comment]
Thanks for testing!

I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org) in 15-30 mins and released in v1.52

**Comment by MikeeI** [comment]
Hi, thanks for documenting this and for fixing the single-hash case with `--hash-type`.

I revisited the remaining multi-hash case against current master
(`92fbc85f108b`, reported as `rclone v1.75.0-DEV`). The original issue mentioned computing all requested
hashes in one read pass, but `lsjson --hash` still reads a local file separately for every hash type.

Current source behavior:

- [`fs/operations/lsjson.go:220-230`](https://github.com/rclone/rclone/blob/92fbc85f108bcb5a990c947025b15216fea6f5a5/fs/operations/lsjson.go#L220-L230)
  calls `Object.Hash` once for each requested hash type.
- [`backend/local/local.go:1204-1224`](https://github.com/rclone/rclone/blob/92fbc85f108bcb5a990c947025b15216fea6f5a5/backend/local/local.go#L1204-L1224)
  opens and reads the file when that individual hash is absent from the object cache.
- Since [PR #7767](https://github.com/rclone/rclone/pull/7767), the local backend supports additional
  hashes by default. On the tested build, plain `--hash` requested 13 hash types.

I measured this with a fully allocated 256 MiB file after warming the page cache:

```text
rclone v1.75.0-DEV
- os/version: ubuntu 24.04 (64 bit)
- os/kernel: 6.8.0-107-generic (x86_64)
- os/type: linux
- os/arch: amd64
- go/version: go1.26.5
- go/linking: dynamic
- go/tags: none
```

Fixture:

```bash
dd if=/dev/zero of=/tmp/rclone-p4.bin bs=1M count=256 status=none
```

Timing without `strace`:

```text
Requested hashes                           elapsed    user     sys
MD5 only                                     0.33s    0.30s   0.04s
MD5, SHA1, SHA256, BLAKE3                    0.72s    0.58s   0.15s
All 13 hashes via --hash                     5.40s    5.04s   0.41s
```

I also counted file operations with:

```bash
strace -f -c -e trace=read,openat \
  rclone lsjson --stat --hash-type md5 /tmp/rclone-p4.bin

strace -f -c -e trace=read,openat \
  rclone lsjson --stat \
  --hash-type md5 --hash-type sha1 --hash-type sha256 --hash-type blake3 \
  /tmp/rclone-p4.bin

strace -f -c -e trace=read,openat \
  rclone lsjson --stat --hash /tmp/rclone-p4.bin
```

Results:

```text
Requested hashes    read calls    openat calls
1                    8,224         19
4                   32,802         22
13                 106,539         31
```

For this file and buffer size, one data pass accounts for 8,192 `read` calls. The counts therefore
correspond almost exactly to one, four, and thirteen complete passes over the same file.

The timing difference also includes the different CPU costs of the hash algorithms, so it does not by
itself predict the speedup from a single-pass implementation. The syscall counts specifically establish
that the file is reread once per requested hash type.

Would it make sense to track the remaining multi-hash case—calculating the requested hash set in one file
pass—or is repeated `--hash-type` with separate passes considered the intended behavior?

I checked all relevant issues, comments, pull requests, and forum threads; this evidence is not already reported.

I am only reporting the finding and am not currently proposing a pull request.

Investigated extensively with GPT-5.6 Sol (xhigh reasoning effort), using [Oh My Pi](https://github.com/can1357/oh-my-pi) as the agent framework.
